File: windows-client/astral_client/protocol_manifest.py
"""Frame classification for the desktop client (feature 044).

Every server->client WS frame type in the committed UI-protocol manifest
(``backend/shared/ui_protocol.json``) is classified here as either

* ``"handled"`` — the client routes it in ``MainWindow._on_message``; or
* ``"ignored"`` — a deliberate, logged drop (the frame carries nothing the
  desktop presents natively; the parity matrix records why).

There is no third state. ``tests/test_protocol_manifest.py`` asserts this table
covers the manifest exactly, so a new server frame type fails the build until it
is classified — never a silent drop (FR-001/FR-002/FR-023).
"""

from __future__ import annotations

HANDLED = "handled"
IGNORED = "ignored"

CLASSIFICATION: dict[str, str] = {
    # bootstrap
    "rote_config": IGNORED,           # natives are full-capability; profile unused
    "chrome_menu": HANDLED,
    "user_preferences": HANDLED,      # theme boot (044)
    "system_config": IGNORED,         # web dashboard payload; desktop uses agent_list
    "agent_list": HANDLED,
    "agent_registered": IGNORED,      # discovery refresh follows via agent_list
    # auth
    "auth_required": HANDLED,
    # canvas / SDUI
    "ui_render": HANDLED,
    "ui_update": IGNORED,             # legacy frame; server no longer targets natives
    "ui_upsert": HANDLED,
    "ui_append": IGNORED,             # legacy frame
    "ui_stream_data": HANDLED,
    # chrome
    "chrome_render": HANDLED,         # web HTML region push -> status notice only
    "chrome_surface": HANDLED,
    # chat lifecycle / progress
    "chat_status": HANDLED,
    "chat_step": HANDLED,
    "chat_created": HANDLED,
    "chat_loaded": HANDLED,
    "chat_deleted": IGNORED,          # cross-tab concern; desktop is single-window
    "history_list": HANDLED,
    "user_message_acked": HANDLED,
    "task_started": HANDLED,
    "task_completed": HANDLED,
    "tool_progress": HANDLED,
    "workspace_timeline_mode": HANDLED,
    "heartbeat": IGNORED,             # transport keepalive
    # streaming
    "stream_subscribed": HANDLED,
    "stream_unsubscribed": HANDLED,
    "stream_list": IGNORED,           # no desktop surface enumerates streams
    "stream_data": HANDLED,
    "stream_error": HANDLED,
    # workspace component verbs (web workspace acks; desktop canvas is ui_* driven)
    "component_saved": IGNORED,
    "component_save_error": IGNORED,
    "saved_components_list": IGNORED,
    "component_deleted": IGNORED,
    "combine_status": IGNORED,
    "combine_error": IGNORED,
    "components_combined": IGNORED,
    "components_condensed": IGNORED,
    # permissions (capability lives in the native Agents dialog via agent_list)
    "agent_permissions": IGNORED,
    "agent_permissions_updated": IGNORED,
    # llm (desktop uses the LLM settings surface round-trip)
    "llm_config_ack": IGNORED,
    "llm_usage_report": IGNORED,
    # audit (desktop fetches audit via REST)
    "audit_append": IGNORED,
    # creation (draft cards carry state in-chat)
    "agent_creation_progress": IGNORED,
    # scheduler notifications + errors (044)
    "notification": HANDLED,
    "error": HANDLED,
}


def is_handled(frame_type: str) -> bool:
    return CLASSIFICATION.get(frame_type) == HANDLED


def is_classified(frame_type: str) -> bool:
    return frame_type in CLASSIFICATION

File: android-client/core/bin/main/com/kyopenscience/astral/core/protocol/ProtocolManifest.kt
package com.kyopenscience.astral.core.protocol

/**
 * Frame classification for the Android client (feature 044).
 *
 * Every server->client WS frame type in the committed UI-protocol manifest
 * (`backend/shared/ui_protocol.json`) is classified as either [HANDLED] (the
 * reducer consumes it) or [IGNORED] (a deliberate, logged drop recorded in the
 * parity matrix). There is no third state: `ProtocolManifestTest` asserts this
 * table covers the manifest exactly, so a new server frame type fails the build
 * until it is classified — never a silent drop (FR-001/FR-002/FR-023).
 */
object ProtocolManifest {
    const val HANDLED = "handled"
    const val IGNORED = "ignored"

    val classification: Map<String, String> =
        mapOf(
            // bootstrap
            // rote_config: natives are full-capability; profile unused
            "rote_config" to IGNORED,
            "chrome_menu" to HANDLED,
            // user_preferences: theme boot (044)
            "user_preferences" to HANDLED,
            // system_config: web dashboard payload; app uses agent_list
            "system_config" to IGNORED,
            "agent_list" to HANDLED,
            "agent_registered" to IGNORED,
            // auth
            "auth_required" to HANDLED,
            // canvas / SDUI
            "ui_render" to HANDLED,
            // ui_update: legacy frame; server no longer targets natives
            "ui_update" to IGNORED,
            "ui_upsert" to HANDLED,
            // ui_append: legacy frame
            "ui_append" to IGNORED,
            "ui_stream_data" to HANDLED,
            // chrome
            // chrome_render: web HTML region push; native gets chrome_surface
            "chrome_render" to IGNORED,
            "chrome_surface" to HANDLED,
            // chat lifecycle / progress
            "chat_status" to HANDLED,
            "chat_step" to HANDLED,
            "chat_created" to HANDLED,
            "chat_loaded" to HANDLED,
            // chat_deleted: cross-tab concern; app is single-window
            "chat_deleted" to IGNORED,
            "history_list" to HANDLED,
            "user_message_acked" to HANDLED,
            "task_started" to HANDLED,
            "task_completed" to HANDLED,
            "tool_progress" to HANDLED,
            "workspace_timeline_mode" to HANDLED,
            // heartbeat: transport keepalive
            "heartbeat" to IGNORED,
            // streaming
            "stream_subscribed" to HANDLED,
            "stream_unsubscribed" to HANDLED,
            // stream_list: no app surface enumerates streams
            "stream_list" to IGNORED,
            "stream_data" to HANDLED,
            "stream_error" to HANDLED,
            // workspace component verbs (web workspace acks; app canvas is ui_* driven)
            "component_saved" to IGNORED,
            "component_save_error" to IGNORED,
            "saved_components_list" to IGNORED,
            "component_deleted" to IGNORED,
            "combine_status" to IGNORED,
            "combine_error" to IGNORED,
            "components_combined" to IGNORED,
            "components_condensed" to IGNORED,
            // permissions (capability lives in the native Agents screen)
            "agent_permissions" to IGNORED,
            "agent_permissions_updated" to IGNORED,
            // llm (app uses the LLM settings surface round-trip)
            "llm_config_ack" to IGNORED,
            "llm_usage_report" to IGNORED,
            // audit (app fetches audit via REST)
            "audit_append" to IGNORED,
            // creation (draft cards carry state in-chat)
            "agent_creation_progress" to IGNORED,
            // scheduler notifications + errors (044)
            "notification" to HANDLED,
            "error" to HANDLED,
        )

    fun isHandled(frameType: String): Boolean = classification[frameType] == HANDLED

    fun isClassified(frameType: String): Boolean = frameType in classification
}
